include/pros/serial.h

```diff
@@ -0,0 +1,247 @@
+ *
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t serial_enable(uint8_t port);
+
+/**
+ * Sets the baudrate for the serial port to operate at.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * EINVAL - The given value is not within the range of V5 ports (1-21).
+ * EACCES - Another resource is currently trying to access the port.
+ *
+ * \param port
+ *        The V5 port number from 1-21
+ * \param baudrate
+ *        The baudrate to operate at
+ *
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t serial_set_baudrate(uint8_t port, int32_t baudrate);
+
+/**
+ * Clears the internal input and output FIFO buffers.
+ *
+ * This can be useful to reset state and remove old, potentially unneeded data
+ * from the input FIFO buffer or to cancel sending any data in the output FIFO
+ * buffer.
+ *
+ * \note This function does not cause the data in the output buffer to be
+ * written, it simply clears the internal buffers. Unlike stdout, generic
+ * serial does not use buffered IO (the FIFO buffers are written as soon
+ * as possible).
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * EINVAL - The given value is not within the range of V5 ports (1-21).
+ * EACCES - Another resource is currently trying to access the port.
+ *
+ * \param port
+ *        The V5 port number from 1-21
+ *
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t serial_flush(uint8_t port);
+
+/**
+ * Returns the number of bytes available to be read in the the port's FIFO
+ * input buffer.
+ *
+ * \note This function does not actually read any bytes, is simply returns the
+ * number of bytes available to be read.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * EINVAL - The given value is not within the range of V5 ports (1-21).
+ * EACCES - Another resource is currently trying to access the port.
+ *
+ * \param port
+ *        The V5 port number from 1-21
+ *
+ * \return The number of bytes avaliable to be read or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t serial_read_avail(uint8_t port);
+
+/**
+ * Returns the number of bytes free in the port's FIFO output buffer.
+ *
+ * \note This function does not actually write any bytes, is simply returns the
+ * number of bytes free in the port's buffer.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * EINVAL - The given value is not within the range of V5 ports (1-21).
+ * EACCES - Another resource is currently trying to access the port.
+ *
+ * \param port
+ *        The V5 port number from 1-21
+ *
+ * \return The number of bytes free or PROS_ERR if the operation failed,
+ * setting errno.
+ */
+int32_t serial_write_free(uint8_t port);
```

Comment:
This function seems a bit confusing to me at first, perhaps the call should be renamed to `serial_get_num_free` or similar? I'm not a fan of having write as the action in this function since no writing occurs.

Reply:
Would `serial_get_write_free` work? I'd like to keep write (or similar) in there just to specify it's returning data about the write buffer. If we're renaming `serial_write_free`, `serial_read_avail` should probably be changed to match.

Reply:
`serial_get_write_free` and `serial_get_ready_avail` sound like good names to me 👍🏻